### PR TITLE
Implementation overhaul

### PR DESCRIPTION
**Pull request by ithinuel**
This update :
- adds support for hardware flow control 
- adds support for error management and safety
- adds support for `NetworkInterface::get_default_instance()`
- fixes buffer overflow and memory leak issues

### CODE REVIEW

**Review by VeijoPesonen — APPROVED** [review]

**Review by VeijoPesonen — APPROVED** [review]
```
mbedgt: test case report:
| target       | platform_name | test suite              | test case                             | passed | failed | result  | elapsed_time (sec) |
|--------------|---------------|-------------------------|---------------------------------------|--------|--------|---------|--------------------|
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS                      | 1      | 0      | OK      | 0.28               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_CACHE                | 1      | 0      | OK      | 0.58               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_CANCEL               | 1      | 0      | OK      | 5.34               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_EXTERNAL_EVENT_QUEUE | 1      | 0      | OK      | 3.22               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_INVALID_HOST         | 1      | 0      | OK      | 1.21               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_NON_ASYNC_AND_ASYNC  | 1      | 0      | OK      | 0.85               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_SIMULTANEOUS         | 1      | 0      | OK      | 0.92               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_SIMULTANEOUS_CACHE   | 1      | 0      | OK      | 0.87               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_SIMULTANEOUS_REPEAT  | 1      | 0      | OK      | 5.23               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | ASYNCHRONOUS_DNS_TIMEOUTS             | 1      | 0      | OK      | 2.79               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | SYNCHRONOUS_DNS                       | 1      | 0      | OK      | 0.1                |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | SYNCHRONOUS_DNS_INVALID               | 1      | 0      | OK      | 1.23               |
| K66F-GCC_ARM | K66F          | tests-netsocket-dns     | SYNCHRONOUS_DNS_MULTIPLE              | 1      | 0      | OK      | 0.54               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_BIND_ADDRESS                | 1      | 0      | OK      | 0.08               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_BIND_ADDRESS_INVALID        | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_BIND_ADDRESS_NULL           | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_BIND_ADDRESS_PORT           | 1      | 0      | OK      | 0.1                |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_BIND_PORT                   | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_BIND_PORT_FAIL              | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_BIND_UNOPENED               | 1      | 0      | OK      | 0.05               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_BIND_WRONG_TYPE             | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_CONNECT_INVALID             | 1      | 0      | OK      | 0.15               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_ECHOTEST                    | 1      | 0      | OK      | 3.58               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_ECHOTEST_BURST              | 1      | 0      | OK      | 7.43               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_ECHOTEST_BURST_NONBLOCK     | 1      | 0      | OK      | 7.45               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_ECHOTEST_NONBLOCK           | 1      | 0      | OK      | 4.16               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_ENDPOINT_CLOSE              | 1      | 0      | OK      | 0.75               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_OPEN_CLOSE_REPEAT           | 1      | 0      | OK      | 0.06               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_OPEN_DESTRUCT               | 1      | 0      | OK      | 0.06               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_OPEN_LIMIT                  | 1      | 0      | OK      | 0.22               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_OPEN_TWICE                  | 1      | 0      | OK      | 0.04               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_RECV_100K                   | 1      | 0      | OK      | 11.19              |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_RECV_100K_NONBLOCK          | 1      | 0      | OK      | 11.95              |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_RECV_TIMEOUT                | 1      | 0      | OK      | 1.44               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_SEND_REPEAT                 | 1      | 0      | OK      | 0.66               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_SEND_TIMEOUT                | 1      | 0      | OK      | 0.24               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_SETSOCKOPT_KEEPALIVE_VALID  | 1      | 0      | OK      | 0.12               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tcp     | TCPSOCKET_THREAD_PER_SOCKET_SAFETY    | 1      | 0      | OK      | 3.29               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_CONNECT_INVALID             | 1      | 0      | OK      | 1.58               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_ECHOTEST                    | 1      | 0      | OK      | 5.16               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_ECHOTEST_BURST              | 1      | 0      | OK      | 8.59               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_ECHOTEST_BURST_NONBLOCK     | 1      | 0      | OK      | 8.62               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_ECHOTEST_NONBLOCK           | 1      | 0      | OK      | 5.13               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_ENDPOINT_CLOSE              | 1      | 0      | OK      | 1.61               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_HANDSHAKE_INVALID           | 1      | 0      | OK      | 0.65               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_NO_CERT                     | 1      | 0      | OK      | 0.58               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_OPEN_DESTRUCT               | 1      | 0      | OK      | 0.07               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_OPEN_LIMIT                  | 1      | 0      | OK      | 0.21               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_OPEN_TWICE                  | 1      | 0      | OK      | 0.04               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_RECV_TIMEOUT                | 1      | 0      | OK      | 2.21               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_SEND_CLOSED                 | 1      | 0      | OK      | 1.6                |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_SEND_REPEAT                 | 1      | 0      | OK      | 2.42               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_SEND_TIMEOUT                | 1      | 0      | OK      | 1.71               |
| K66F-GCC_ARM | K66F          | tests-netsocket-tls     | TLSSOCKET_SEND_UNCONNECTED            | 1      | 0      | OK      | 0.06               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_BIND_ADDRESS                | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_BIND_ADDRESS_INVALID        | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_BIND_ADDRESS_NULL           | 1      | 0      | OK      | 0.1                |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_BIND_ADDRESS_PORT           | 1      | 0      | OK      | 0.1                |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_BIND_PORT                   | 1      | 0      | OK      | 0.08               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_BIND_PORT_FAIL              | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_BIND_UNOPENED               | 1      | 0      | OK      | 0.06               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_BIND_WRONG_TYPE             | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_ECHOTEST                    | 0      | 0      | ERROR   | 0.0                |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_ECHOTEST_BURST              | 0      | 0      | SKIPPED | 0.0                |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_ECHOTEST_BURST_NONBLOCK     | 1      | 0      | OK      | 6.68               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_ECHOTEST_NONBLOCK           | 1      | 0      | OK      | 3.46               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_OPEN_CLOSE_REPEAT           | 1      | 0      | OK      | 0.04               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_OPEN_DESTRUCT               | 1      | 0      | OK      | 0.03               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_OPEN_LIMIT                  | 1      | 0      | OK      | 0.34               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_OPEN_TWICE                  | 1      | 0      | OK      | 0.05               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_RECV_TIMEOUT                | 1      | 0      | OK      | 1.2                |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_SENDTO_INVALID              | 1      | 0      | OK      | 0.06               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_SENDTO_REPEAT               | 1      | 0      | OK      | 0.55               |
| K66F-GCC_ARM | K66F          | tests-netsocket-udp     | UDPSOCKET_SENDTO_TIMEOUT              | 1      | 0      | OK      | 0.15               |
| K66F-GCC_ARM | K66F          | tests-network-interface | NETWORKINTERFACE_CONN_DISC_REPEAT     | 1      | 0      | OK      | 20.21              |
| K66F-GCC_ARM | K66F          | tests-network-interface | NETWORKINTERFACE_STATUS               | 1      | 0      | OK      | 26.66              |
| K66F-GCC_ARM | K66F          | tests-network-interface | NETWORKINTERFACE_STATUS_GET           | 1      | 0      | OK      | 24.24              |
| K66F-GCC_ARM | K66F          | tests-network-interface | NETWORKINTERFACE_STATUS_NONBLOCK      | 1      | 0      | OK      | 19.91              |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT                          | 1      | 0      | OK      | 11.33              |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-DISCONNECT-REPEAT        | 1      | 0      | OK      | 37.94              |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-NOCREDENTIALS            | 1      | 0      | OK      | 0.06               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-PARAMS-CHANNEL           | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-PARAMS-CHANNEL-FAIL      | 1      | 0      | OK      | 0.11               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-PARAMS-NULL              | 1      | 0      | OK      | 0.06               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-PARAMS-VALID-SECURE      | 1      | 0      | OK      | 5.17               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-PARAMS-VALID-UNSECURE    | 1      | 0      | OK      | 7.59               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-SECURE                   | 1      | 0      | OK      | 3.87               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONNECT-SECURE-FAIL              | 1      | 0      | OK      | 12.68              |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-CONSTRUCTOR                      | 1      | 0      | OK      | 0.05               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-GET-RSSI                         | 1      | 0      | OK      | 3.91               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-SCAN                             | 1      | 0      | OK      | 1.53               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-SCAN-NULL                        | 1      | 0      | OK      | 6.15               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-SET-CHANNEL                      | 1      | 0      | OK      | 0.09               |
| K66F-GCC_ARM | K66F          | tests-network-wifi      | WIFI-SET-CREDENTIAL                   | 1      | 0      | OK      | 0.05               |
mbedgt: test case results: 1 SKIPPED / 92 OK / 1 ERROR
```

**Comment by VeijoPesonen** [comment]
@wpgnss would you please review this PR.

**Review by wpgnss — APPROVED** [review]

File: WizFi310/WizFi310.h [hunk 1]
```diff
@@ -36,223 +36,412 @@
...
-    */
-    bool dhcp(bool enabled);
+     * Creates a new WizFir310 driver.
+     * @param tx    Tx pin name.
+     * @param rx    Rx pin name.
+     * @param rts   Rts pin name.
+     * @param cts   Cts pin name.
+     * @param rst   Rst pin name.
+     */
+    WizFi310(PinName tx, PinName rx, PinName rtx = NC, PinName cts = NC, PinName rst = NC);
```
Comment (JanneKiiskila) [inline WizFi310/WizFi310.h]: Should the `rtx` pin be called actually `rts`?

**Comment by JanneKiiskila** [comment]
Using NUCLEO L4R5ZI - it will crash with Pinmap issues.
If you do not specify CTS, RTS, RST - you get this;

```
++ MbedOS Error Info ++
Error Status: 0x80FF0144 Code: 324 Module: 255
Error Message: Assertion failed: uart_rts != (UARTName)NC
Location: 0x80295B5
File: ./mbed-os/targets/TARGET_STM/TARGET_STM32L4/serial_device.c+706
Error Value: 0x0
Current Thread: wizfi310_driver  Id: 0x200040FC Entry: 0x802CBAD StackSize: 0x800 StackMem: 0x20007 
For more info, visit: https://mbed.com/s/error?error=0x80FF0144&tgt=NUCLEO_L4R5ZI
-- MbedOS Error Info --
```

If you specify those, it will crash differently.

```
++ MbedOS Error Info ++
Error Status: 0x80010130 Code: 304 Module: 1
Error Message: pinmap not found for peripheral
Location: 0x8028AA1
Error Value: 0xFFFFFFFF
Current Thread: wizfi310_driver  Id: 0x200040FC Entry: 0x802CBAD StackSize: 0x800 StackMem: 0x20007 
For more info, visit: https://mbed.com/s/error?error=0x80010130&tgt=NUCLEO_L4R5ZI
-- MbedOS Error Info --
```

With the following definitions:

```
            "wizfi310.provide-default"                  : true, 
            "wizfi310.tx"                               : "D1",
            "wizfi310.rx"                               : "D0",
            "wizfi310.cts"                              : "D6",
            "wizfi310.rts"                              : "D7",
            "wizfi310.rst"                              : "D9",
```

Those seem to be the deault Arduino header pins the board attaches to (ref: https://www.wiznet.io/product-item/wizfi310-shield/).

**Comment by JanneKiiskila** [comment]
This can be easily duplicated by using the `mbed-os-example-wifi` repo.
Using wizfi310-driver master the example will pass with the D1, D0, D6, D7 definitions.

File: WizFi310/WizFi310.h (hunk 1 above)
Comment (ithinuel) [reply WizFi310/WizFi310.h]: Indeed, I fixed that typo. Thanks!

**Comment by ithinuel** [comment]
@JanneKiiskila I made a small change that should render RTS/CTS optional.

However HW flow control is the only way to ensure data integrity. This is especially important for stress tests that require either high bandwidth or a reliable flow control mechanism.

**Comment by JanneKiiskila** [comment]
Fully agree.

**Comment by ithinuel** [comment]
This PR is more than 1 year old and was not merged while it was not conflicting. Closing.

**Comment by JanneKiiskila** [comment]
Were the issues fixed, so that it could have been merged?

**Comment by ithinuel** [comment]
Back in November 2019 I think so. There's now several merge conflicts that I won't have time to resolve.
The source branch on my repo is still public if anyone wants to take it over.

**Pull request closed without merging**